Flash erase does not work on STM32G473CB chip

Flash erase does not work on STM32G473CB chip with 128 KB flash memory with 384 KB hole between 2 banks. It does work on STM32G474RE chip with 512 KB flash and there is no memory hole between 2 banks.
I tried to erase last 2 pages (2 * 2028 KB) in the flash and it failed with the following messages:
 
.000122 [TRACE] Erasing from 0x804f000 to 0x8050000 (<crate> └─ <invalid location: defmt frame-index: 14>:0)
0.000152 [TRACE] Erasing sector: FlashSector { bank: Bank2, index_in_bank: 30, start: 134541312, size: 2048 } (<crate> └─ <invalid location: defmt frame-index: 15>:0)
0.022216 [TRACE] Erasing sector: FlashSector { bank: Bank2, index_in_bank: 31, start: 134543360, size: 2048 } (<crate> └─ <invalid location: defmt frame-index: 15>:0)

I copied the example test file "flash_async.rs" from "../examples/stm32g474/" into "../examples/stm32g4/", set chip to stm32g473cb,  and ran a flash erase test from your PR--"fix-stm32g473cb-flash" on STM32 microcontroller STM32G473CBT6 and it failed.

 cargo run --bin flash_async

0.000152 [INFO ] Hello Flash! (flash_async src/bin/flash_async.rs:19)
0.000915 [INFO ] Testing offset: 0x1F000, size: 0x1000 (flash_async src/bin/flash_async.rs:48)
0.001953 [INFO ] Reading... (flash_async src/bin/flash_async.rs:50)
0.002624 [INFO ] Read: [ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff] (flash_async src/bin/flash_async.rs:53)
0.006011 [INFO ] Erasing... (flash_async src/bin/flash_async.rs:55)
0.006805 [ERROR] panicked at 'unwrap failed: f.erase(offset, offset + size).await'
error: `Miss` (flash_async src/bin/flash_async.rs:56)
0.007659 [ERROR] panicked at C:\Users\E5000496\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\defmt-1.1.0\src\lib.rs:385:5:
explicit panic (panic_probe panic-probe-1.0.0/src/lib.rs:104)
Firmware exited unexpectedly: Multiple


The source code change from 
    test_flash(&mut f, 256 * 1024, 4 * 1024).await;     --> erase first 4 KB flash on bank 2 for STM32G474RE
into 
    test_flash(&mut f, 124 * 1024, 4 * 1024).await;     --> erase last 4 KB flash on bank 2 for STM32G473CB

--- a/embassy-stm32/src/flash/asynch.rs
+++ b/embassy-stm32/src/flash/asynch.rs
@@ -7,7 +7,7 @@ use embassy_sync::mutex::Mutex;
 
 use super::{
     Async, Error, FLASH_BASE, FLASH_SIZE, Flash, FlashLayout, WRITE_SIZE, blocking_read, ensure_sector_aligned, family,
-    get_flash_regions, get_sector,
+    flash_addressable_size, get_flash_regions, get_sector,
 };
 use crate::interrupt::InterruptExt;
 use crate::peripherals::FLASH;
@@ -46,7 +46,7 @@ impl<'d> Flash<'d, Async> {
     /// NOTE: `offset` is an offset from the flash start, NOT an absolute address.
     /// For example, to write address `0x0800_1234` you have to use offset `0x1234`.
     pub async fn write(&mut self, offset: u32, bytes: &[u8]) -> Result<(), Error> {
-        unsafe { write_chunked(FLASH_BASE as u32, FLASH_SIZE as u32, offset, bytes).await }
+        unsafe { write_chunked(FLASH_BASE as u32, flash_addressable_size(), offset, bytes).await }
     }
 
     /// Async erase.

--- a/embassy-stm32/src/flash/common.rs
+++ b/embassy-stm32/src/flash/common.rs
@@ -44,7 +44,7 @@ impl<'d, MODE> Flash<'d, MODE> {
     /// NOTE: `offset` is an offset from the flash start, NOT an absolute address.
     /// For example, to read address `0x0800_1234` you have to use offset `0x1234`.
     pub fn blocking_read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Error> {
-        blocking_read(FLASH_BASE as u32, FLASH_SIZE as u32, offset, bytes)
+        blocking_read(FLASH_BASE as u32, flash_addressable_size(), offset, bytes)
     }
 
     /// Blocking write.
@@ -55,7 +55,7 @@ impl<'d, MODE> Flash<'d, MODE> {
         unsafe {
             blocking_write(
                 FLASH_BASE as u32,
-                FLASH_SIZE as u32,
+                flash_addressable_size(),
                 offset,
                 bytes,
                 write_chunk_unlocked,
@@ -175,6 +175,19 @@ pub(super) unsafe fn erase_sector_unlocked(sector: &FlashSector) -> Result<(), E
 
 pub(super) unsafe fn erase_sector_with_critical_section(sector: &FlashSector) -> Result<(), Error> {
     critical_section::with(|_| erase_sector_unlocked(sector))
+}
+
+/// Size of the address range spanned by all flash regions, from the start of the first
+/// region to the end of the last region.
+///
+/// This can be larger than `FLASH_SIZE` (the sum of region sizes) on chips where the banks
+/// are not contiguous, e.g. STM32G473CB has a 192KB gap between bank 1 and bank 2. Bounding
+/// offsets by `FLASH_SIZE` in that case would make the later bank unreachable even though
+/// `get_sector` can resolve its addresses just fine.
+pub(super) fn flash_addressable_size() -> u32 {
+    let regions = get_flash_regions();
+    let end = regions.iter().map(|r| r.end()).max().unwrap();
+    end - FLASH_BASE as u32
 }
 
 pub(super) fn get_sector(address: u32, regions: &[&FlashRegion]) -> Result<FlashSector, Error> {

